File: src/lws/providers/stepfunctions/engine.py
"""Step Functions execution engine.

Implements the recursive state walker that processes state machine definitions.
Handles all state types, error handling (Retry/Catch), and execution tracking.
"""

from __future__ import annotations

import asyncio
import logging
import time
import uuid
from dataclasses import dataclass, field
from datetime import UTC, datetime
from enum import Enum
from typing import Any, Protocol

from lws.providers.stepfunctions.asl_parser import (
    CatchConfig,
    ChoiceState,
    FailState,
    MapState,
    ParallelState,
    PassState,
    RetryConfig,
    StateMachineDefinition,
    SucceedState,
    TaskState,
    WaitState,
)
from lws.providers.stepfunctions.choice_evaluator import evaluate_choice_rules
from lws.providers.stepfunctions.path_utils import (
    apply_context_parameters,
    apply_input_path,
    apply_output_path,
    apply_parameters,
    apply_result_path,
    resolve_path,
)

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Execution tracking (P2-15)
# ---------------------------------------------------------------------------


class ExecutionStatus(Enum):
    """Status of a state machine execution."""

    RUNNING = "RUNNING"
    SUCCEEDED = "SUCCEEDED"
    FAILED = "FAILED"
    TIMED_OUT = "TIMED_OUT"
    ABORTED = "ABORTED"


@dataclass
class StateTransition:
    """Record of a single state transition during execution."""

    state_name: str
    state_type: str
    timestamp: float
    input_data: Any = None
    output_data: Any = None
    error: str | None = None
    cause: str | None = None


@dataclass
class ExecutionHistory:
    """Complete history of a state machine execution."""

    execution_arn: str
    state_machine_name: str
    status: ExecutionStatus = ExecutionStatus.RUNNING
    start_time: float = 0.0
    end_time: float | None = None
    input_data: Any = None
    output_data: Any = None
    error: str | None = None
    cause: str | None = None
    transitions: list[StateTransition] = field(default_factory=list)


# ---------------------------------------------------------------------------
# Custom exceptions
# ---------------------------------------------------------------------------


class StatesError(Exception):
    """Base exception for state machine execution errors."""

    def __init__(self, error: str, cause: str | None = None) -> None:
        super().__init__(error)
        self.error = error
        self.cause = cause


class StatesTaskFailed(StatesError):
    """Raised when a task state fails."""


class StatesTimeout(StatesError):
    """Raised when a state times out."""


# ---------------------------------------------------------------------------
# Compute interface protocol
# ---------------------------------------------------------------------------


class ComputeInvoker(Protocol):
    """Protocol for invoking compute functions."""

    async def invoke_function(self, resource_arn: str, payload: Any) -> Any:
        """Invoke a compute function by resource ARN with the given payload."""


# ---------------------------------------------------------------------------
# Execution engine
# ---------------------------------------------------------------------------


class ExecutionEngine:
    """Executes a state machine definition step by step.

    Parameters
    ----------
    definition:
        The parsed state machine definition.
    compute:
        Optional compute invoker for Task states.
    max_wait_seconds:
        Maximum wait time compression for local development.
    """

    def __init__(
        self,
        definition: StateMachineDefinition,
        compute: ComputeInvoker | None = None,
        max_wait_seconds: float = 5.0,
    ) -> None:
        self._definition = definition
        self._compute = compute
        self._max_wait_seconds = max_wait_seconds

    async def execute(
        self,
        input_data: Any = None,
        execution_arn: str | None = None,
        state_machine_name: str = "local-state-machine",
    ) -> ExecutionHistory:
        """Execute the state machine with the given input."""
        if execution_arn is None:
            uid = uuid.uuid4()
            execution_arn = (
                f"arn:aws:states:us-east-1:000000000000:execution:{state_machine_name}:{uid}"
            )

        history = ExecutionHistory(
            execution_arn=execution_arn,
            state_machine_name=state_machine_name,
            start_time=time.time(),
            input_data=input_data,
        )

        try:
            result = await self._run_state_machine(input_data, history)
            _mark_succeeded(history, result)
        except StatesError as exc:
            _mark_failed(history, exc.error, exc.cause)
        except Exception as exc:
            _mark_failed(history, "States.Runtime", str(exc))

        return history

    async def _run_state_machine(
        self,
        input_data: Any,
        history: ExecutionHistory,
    ) -> Any:
        """Walk through states from StartAt until terminal state."""
        current_state_name = self._definition.start_at
        current_data = input_data

        while current_state_name is not None:
            state = self._definition.states.get(current_state_name)
            if state is None:
                raise StatesError("States.Runtime", f"State not found: {current_state_name}")

            transition = StateTransition(
                state_name=current_state_name,
                state_type=type(state).__name__,
                timestamp=time.time(),
                input_data=current_data,
            )
            history.transitions.append(transition)

            current_data, current_state_name = await self._execute_state(
                state, current_data, transition
            )
            transition.output_data = current_data

        return current_data

    async def _execute_state(
        self,
        state: Any,
        input_data: Any,
        _transition: StateTransition,
    ) -> tuple[Any, str | None]:
        """Execute a single state and return (output, next_state_name)."""
        handlers = {
            PassState: self._execute_pass,
            TaskState: self._execute_task,
            ChoiceState: self._execute_choice,
            WaitState: self._execute_wait,
            SucceedState: self._execute_succeed,
            FailState: self._execute_fail,
            ParallelState: self._execute_parallel,
            MapState: self._execute_map,
        }
        handler = handlers.get(type(state))
        if handler is None:
            raise StatesError("States.Runtime", f"Unsupported state type: {type(state).__name__}")
        return await handler(state, input_data)

    # -------------------------------------------------------------------
    # Pass state
    # -------------------------------------------------------------------

    async def _execute_pass(self, state: PassState, input_data: Any) -> tuple[Any, str | None]:
        """Execute a Pass state."""
        effective_input = apply_input_path(input_data, state.input_path)
        if state.parameters:
            effective_input = apply_parameters(state.parameters, effective_input)
        result = state.result if state.result is not None else effective_input
        output = apply_result_path(effective_input, result, state.result_path)
        output = apply_output_path(output, state.output_path)
        return output, _next_or_none(state.next_state, state.end)

    # -------------------------------------------------------------------
    # Task state (P2-09)
    # -------------------------------------------------------------------

    async def _execute_task(self, state: TaskState, input_data: Any) -> tuple[Any, str | None]:
        """Execute a Task state with retry/catch support."""
        effective_input = _prepare_task_input(state, input_data)
        try:
            result = await self._invoke_with_retry(state, effective_input)
            return _apply_task_output(state, input_data, result)
        except StatesError as exc:
            return _handle_task_catch(state, input_data, exc)

    async def _invoke_with_retry(self, state: TaskState, effective_input: Any) -> Any:
        """Invoke a task with retry logic."""
        if not state.retry:
            return await self._invoke_task(state, effective_input)

        last_error: StatesError | None = None
        for retry_config in state.retry:
            result_or_error = await self._attempt_retry_block(state, effective_input, retry_config)
            if not isinstance(result_or_error, StatesError):
                return result_or_error
            last_error = result_or_error

        if last_error is not None:
            raise last_error
        return await self._invoke_task(state, effective_input)

    async def _attempt_retry_block(
        self,
        state: TaskState,
        effective_input: Any,
        retry_config: RetryConfig,
    ) -> Any:
        """Attempt a single retry block. Returns result on success or StatesError on failure."""
        last_error: StatesError | None = None
        for attempt in range(retry_config.max_attempts + 1):
            try:
                return await self._invoke_task(state, effective_input)
            except StatesError as exc:
                if not _error_matches(exc.error, retry_config.error_equals):
                    raise
                last_error = exc
                if attempt < retry_config.max_attempts:
                    delay = _calculate_retry_delay(retry_config, attempt)
                    await asyncio.sleep(delay)

        if last_error is not None:
            return last_error
        raise StatesError("States.Runtime", "Unexpected retry state")

    async def _invoke_task(self, state: TaskState, payload: Any) -> Any:
        """Invoke the task resource with optional timeout."""
        if self._compute is None:
            raise StatesError("States.Runtime", "No compute invoker configured")

        if state.timeout_seconds:
            return await self._invoke_with_timeout(state, payload)

        return await self._compute.invoke_function(state.resource, payload)

    async def _invoke_with_timeout(self, state: TaskState, payload: Any) -> Any:
        """Invoke a task with a timeout enforced."""
        try:
            return await asyncio.wait_for(
                self._compute.invoke_function(state.resource, payload),
                timeout=state.timeout_seconds,
            )
        except TimeoutError as exc:
            raise StatesTimeout(
                "States.Timeout",
                f"Task timed out after {state.timeout_seconds}s",
            ) from exc

    # -------------------------------------------------------------------
    # Choice state (P2-10)
    # -------------------------------------------------------------------

    async def _execute_choice(self, state: ChoiceState, input_data: Any) -> tuple[Any, str | None]:
        """Execute a Choice state by evaluating rules."""
        effective_input = apply_input_path(input_data, state.input_path)
        next_state = evaluate_choice_rules(state.choices, effective_input)
        if next_state is None:
            next_state = state.default
        if next_state is None:
            raise StatesError("States.NoChoiceMatched", "No choice rule matched and no Default")
        output = apply_output_path(effective_input, state.output_path)
        return output, next_state

    # -------------------------------------------------------------------
    # Wait state (P2-11)
    # -------------------------------------------------------------------

    async def _execute_wait(self, state: WaitState, input_data: Any) -> tuple[Any, str | None]:
        """Execute a Wait state."""
        effective_input = apply_input_path(input_data, state.input_path)
        wait_seconds = _resolve_wait_seconds(state, effective_input)
        clamped = min(wait_seconds, self._max_wait_seconds)
        if clamped > 0:
            await asyncio.sleep(clamped)
        output = apply_output_path(effective_input, state.output_path)
        return output, _next_or_none(state.next_state, state.end)

    # -------------------------------------------------------------------
    # Succeed / Fail states
    # -------------------------------------------------------------------

    async def _execute_succeed(
        self, state: SucceedState, input_data: Any
    ) -> tuple[Any, str | None]:
        """Execute a Succeed state (terminal)."""
        effective_input = apply_input_path(input_data, state.input_path)
        output = apply_output_path(effective_input, state.output_path)
        return output, None

    async def _execute_fail(self, state: FailState, input_data: Any) -> tuple[Any, str | None]:
        """Execute a Fail state (terminal with error)."""
        raise StatesError(
            error=state.error or "States.Fail",
            cause=state.cause,
        )

    # -------------------------------------------------------------------
    # Parallel state (P2-13)
    # -------------------------------------------------------------------

    async def _execute_parallel(
        self, state: ParallelState, input_data: Any
    ) -> tuple[Any, str | None]:
        """Execute a Parallel state with concurrent branches."""
        effective_input = apply_input_path(input_data, state.input_path)
        try:
            results = await self._run_parallel_branches(state, effective_input)
        except StatesError as exc:
            return _handle_parallel_catch(state, input_data, exc)

        output = _apply_parallel_output(state, input_data, results)
        return output, _next_or_none(state.next_state, state.end)

    async def _run_parallel_branches(self, state: ParallelState, effective_input: Any) -> list[Any]:
        """Run all branches concurrently and collect results."""
        tasks = [self._run_branch(branch, effective_input) for branch in state.branches]
        try:
            results = await asyncio.gather(*tasks, return_exceptions=False)
        except Exception:
            # Cancel remaining tasks on branch failure
            for task in tasks:
                if isinstance(task, asyncio.Task) and not task.done():
                    task.cancel()
            raise
        return list(results)

    async def _run_branch(self, branch: StateMachineDefinition, input_data: Any) -> Any:
        """Run a single branch as a sub-state-machine."""
        sub_engine = ExecutionEngine(
            definition=branch,
            compute=self._compute,
            max_wait_seconds=self._max_wait_seconds,
        )
        history = await sub_engine.execute(input_data)
        if history.status == ExecutionStatus.FAILED:
            raise StatesError(
                error=history.error or "States.BranchFailed",
                cause=history.cause,
            )
        return history.output_data

    # -------------------------------------------------------------------
    # Map state (P2-14)
    # -------------------------------------------------------------------

    async def _execute_map(self, state: MapState, input_data: Any) -> tuple[Any, str | None]:
        """Execute a Map state, iterating over items."""
        effective_input = apply_input_path(input_data, state.input_path)
        items = _resolve_map_items(state, effective_input)

        try:
            results = await self._run_map_iterations(state, effective_input, items)
        except StatesError as exc:
            return _handle_map_catch(state, input_data, exc)

        output = apply_result_path(effective_input, results, state.result_path)
        if state.result_selector:
            output = apply_parameters(state.result_selector, output)
        output = apply_output_path(output, state.output_path)
        return output, _next_or_none(state.next_state, state.end)

    async def _run_map_iterations(
        self,
        state: MapState,
        effective_input: Any,
        items: list,
    ) -> list[Any]:
        """Run Map iterations with optional concurrency limiting."""
        if state.iterator is None:
            raise StatesError("States.Runtime", "Map state has no Iterator")

        semaphore = _create_map_semaphore(state.max_concurrency)

        tasks = [
            self._run_map_item(state, effective_input, item, index, semaphore)
            for index, item in enumerate(items)
        ]
        return list(await asyncio.gather(*tasks))

    async def _run_map_item(
        self,
        state: MapState,
        effective_input: Any,
        item: Any,
        index: int,
        semaphore: asyncio.Semaphore | None,
    ) -> Any:
        """Run a single Map iteration."""
        item_input = _build_map_item_input(state, effective_input, item, index)

        if semaphore:
            async with semaphore:
                return await self._run_branch(state.iterator, item_input)
        return await self._run_branch(state.iterator, item_input)


# ---------------------------------------------------------------------------
# Helper functions (extracted to keep method complexity low)
# ---------------------------------------------------------------------------


def _next_or_none(next_state: str | None, end: bool) -> str | None:
    """Return next state name or None if this is a terminal state."""
    if end:
        return None
    return next_state


def _prepare_task_input(state: TaskState, input_data: Any) -> Any:
    """Prepare the effective input for a task invocation."""
    effective_input = apply_input_path(input_data, state.input_path)
    if state.parameters:
        effective_input = apply_parameters(state.parameters, effective_input)
    return effective_input


def _apply_task_output(
    state: TaskState, original_input: Any, result: Any
) -> tuple[Any, str | None]:
    """Apply ResultSelector, ResultPath, and OutputPath to task result."""
    if state.result_selector:
        result = apply_parameters(state.result_selector, result)
    output = apply_result_path(original_input, result, state.result_path)
    output = apply_output_path(output, state.output_path)
    return output, _next_or_none(state.next_state, state.end)


def _handle_task_catch(
    state: TaskState, input_data: Any, exc: StatesError
) -> tuple[Any, str | None]:
    """Handle a task error using Catch configuration."""
    if state.catch:
        for catch_config in state.catch:
            if _error_matches(exc.error, catch_config.error_equals):
                return _apply_catch(input_data, catch_config, exc)
    raise exc


def _handle_parallel_catch(
    state: ParallelState, input_data: Any, exc: StatesError
) -> tuple[Any, str | None]:
    """Handle a parallel branch error using Catch configuration."""
    if state.catch:
        for catch_config in state.catch:
            if _error_matches(exc.error, catch_config.error_equals):
                return _apply_catch(input_data, catch_config, exc)
    raise exc


def _handle_map_catch(state: MapState, input_data: Any, exc: StatesError) -> tuple[Any, str | None]:
    """Handle a map iteration error using Catch configuration."""
    if state.catch:
        for catch_config in state.catch:
            if _error_matches(exc.error, catch_config.error_equals):
                return _apply_catch(input_data, catch_config, exc)
    raise exc


def _apply_catch(
    input_data: Any, catch_config: CatchConfig, exc: StatesError
) -> tuple[Any, str | None]:
    """Apply catch configuration to produce output and next state."""
    error_info = {"Error": exc.error, "Cause": exc.cause or str(exc)}
    output = apply_result_path(input_data, error_info, catch_config.result_path)
    return output, catch_config.next_state


def _apply_parallel_output(state: ParallelState, original_input: Any, results: list[Any]) -> Any:
    """Apply ResultSelector, ResultPath, and OutputPath to parallel results."""
    result = results
    if state.result_selector:
        result = apply_parameters(state.result_selector, result)
    output = apply_result_path(original_input, result, state.result_path)
    return apply_output_path(output, state.output_path)


def _error_matches(error_name: str, error_equals: list[str]) -> bool:
    """Check if an error name matches any of the error patterns."""
    if "States.ALL" in error_equals:
        return True
    return error_name in error_equals


def _calculate_retry_delay(retry_config: RetryConfig, attempt: int) -> float:
    """Calculate the delay before the next retry attempt."""
    return retry_config.interval_seconds * (retry_config.backoff_rate**attempt)


def _resolve_wait_seconds(state: WaitState, input_data: Any) -> float:
    """Resolve the number of seconds to wait based on state configuration."""
    if state.seconds is not None:
        return float(state.seconds)
    if state.seconds_path is not None:
        val = resolve_path(input_data, state.seconds_path)
        return float(val)
    if state.timestamp is not None:
        return _seconds_until_timestamp(state.timestamp)
    if state.timestamp_path is not None:
        ts = resolve_path(input_data, state.timestamp_path)
        return _seconds_until_timestamp(ts)
    return 0.0


def _seconds_until_timestamp(timestamp: str) -> float:
    """Calculate seconds until an ISO 8601 timestamp."""
    target = datetime.fromisoformat(timestamp.replace("Z", "+00:00"))
    now = datetime.now(UTC)
    diff = (target - now).total_seconds()
    return max(0.0, diff)


def _resolve_map_items(state: MapState, effective_input: Any) -> list:
    """Resolve the items to iterate over in a Map state."""
    if state.items_path and state.items_path != "$":
        items = resolve_path(effective_input, state.items_path)
    else:
        items = effective_input
    if not isinstance(items, list):
        raise StatesError("States.Runtime", "Map state items must be a list")
    return items


def _build_map_item_input(
    state: MapState,
    effective_input: Any,
    item: Any,
    index: int,
) -> Any:
    """Build the input for a single Map iteration."""
    if state.parameters:
        context = {"Map": {"Item": {"Value": item, "Index": index}}}
        return apply_context_parameters(state.parameters, effective_input, context)
    return item


def _create_map_semaphore(max_concurrency: int) -> asyncio.Semaphore | None:
    """Create a semaphore for Map concurrency limiting, or None for unlimited."""
    if max_concurrency > 0:
        return asyncio.Semaphore(max_concurrency)
    return None


def _mark_succeeded(history: ExecutionHistory, result: Any) -> None:
    """Mark an execution as succeeded."""
    history.status = ExecutionStatus.SUCCEEDED
    history.output_data = result
    history.end_time = time.time()


def _mark_failed(history: ExecutionHistory, error: str, cause: str | None) -> None:
    """Mark an execution as failed."""
    history.status = ExecutionStatus.FAILED
    history.error = error
    history.cause = cause
    history.end_time = time.time()

File: lang/typescript/core/src/providers/stepfunctions/engine.ts
/** Step Functions ASL interpreter. */

import { v4 as uuidv4 } from "uuid";

export interface StateMachineDefinition {
  Comment?: string;
  StartAt: string;
  States: Record<string, AslState>;
  TimeoutSeconds?: number;
}

export type AslState =
  | PassState
  | TaskState
  | ChoiceState
  | WaitState
  | SucceedState
  | FailState
  | ParallelState
  | MapState;

export interface BaseState {
  Type: string;
  Comment?: string;
  InputPath?: string;
  OutputPath?: string;
  ResultPath?: string;
  Parameters?: Record<string, unknown>;
  Next?: string;
  End?: boolean;
  Retry?: RetryConfig[];
  Catch?: CatchConfig[];
}

export interface PassState extends BaseState {
  Type: "Pass";
  Result?: unknown;
}

export interface TaskState extends BaseState {
  Type: "Task";
  Resource: string;
  TimeoutSeconds?: number;
  HeartbeatSeconds?: number;
}

export interface ChoiceState {
  Type: "Choice";
  Comment?: string;
  InputPath?: string;
  OutputPath?: string;
  Choices: ChoiceRule[];
  Default?: string;
}

export interface ChoiceRule {
  Next: string;
  Variable?: string;
  StringEquals?: string;
  StringEqualsPath?: string;
  StringLessThan?: string;
  StringGreaterThan?: string;
  StringLessThanOrEquals?: string;
  StringGreaterThanOrEquals?: string;
  StringMatches?: string;
  NumericEquals?: number;
  NumericLessThan?: number;
  NumericGreaterThan?: number;
  NumericLessThanOrEquals?: number;
  NumericGreaterThanOrEquals?: number;
  BooleanEquals?: boolean;
  IsNull?: boolean;
  IsPresent?: boolean;
  IsString?: boolean;
  IsNumeric?: boolean;
  IsBoolean?: boolean;
  And?: ChoiceRule[];
  Or?: ChoiceRule[];
  Not?: ChoiceRule;
}

export interface WaitState extends BaseState {
  Type: "Wait";
  Seconds?: number;
  SecondsPath?: string;
  Timestamp?: string;
  TimestampPath?: string;
}

export interface SucceedState {
  Type: "Succeed";
  Comment?: string;
  InputPath?: string;
  OutputPath?: string;
}

export interface FailState {
  Type: "Fail";
  Comment?: string;
  Error?: string;
  Cause?: string;
}

export interface ParallelState extends BaseState {
  Type: "Parallel";
  Branches: Array<{ StartAt: string; States: Record<string, AslState> }>;
}

export interface MapState extends BaseState {
  Type: "Map";
  ItemsPath?: string;
  Iterator: { StartAt: string; States: Record<string, AslState> };
  MaxConcurrency?: number;
}

export interface RetryConfig {
  ErrorEquals: string[];
  IntervalSeconds?: number;
  MaxAttempts?: number;
  BackoffRate?: number;
}

export interface CatchConfig {
  ErrorEquals: string[];
  Next: string;
  ResultPath?: string;
}

export interface Execution {
  executionArn: string;
  stateMachineArn: string;
  name: string;
  status: "RUNNING" | "SUCCEEDED" | "FAILED" | "TIMED_OUT" | "ABORTED";
  startDate: number;
  stopDate?: number;
  input: string;
  output?: string;
  error?: string;
  cause?: string;
}

class StatesError extends Error {
  constructor(
    public readonly errorCode: string,
    public readonly cause?: string,
  ) {
    super(errorCode);
  }
}

// ─── Path utilities ──────────────────────────────────────────────────────────

function resolvePath(data: unknown, path: string | undefined): unknown {
  if (!path || path === "$") return data;
  if (!path.startsWith("$.")) return data;
  const keys = path.slice(2).split(".");
  let cur = data as Record<string, unknown>;
  for (const key of keys) {
    if (cur === null || cur === undefined || typeof cur !== "object") return undefined;
    cur = (cur as Record<string, unknown>)[key] as Record<string, unknown>;
  }
  return cur;
}

function setPath(data: unknown, path: string | undefined, value: unknown): unknown {
  if (!path || path === "$") return value;
  if (!path.startsWith("$.")) return data;
  const keys = path.slice(2).split(".");
  const result =
    typeof data === "object" && data !== null ? { ...(data as Record<string, unknown>) } : {};
  let cur: Record<string, unknown> = result;
  for (let i = 0; i < keys.length - 1; i++) {
    const key = keys[i];
    if (typeof cur[key] !== "object" || cur[key] === null) cur[key] = {};
    cur[key] = { ...(cur[key] as Record<string, unknown>) };
    cur = cur[key] as Record<string, unknown>;
  }
  cur[keys[keys.length - 1]] = value;
  return result;
}

function applyInputPath(data: unknown, inputPath: string | undefined): unknown {
  if (inputPath === null) return {};
  return resolvePath(data, inputPath ?? "$");
}

function applyResultPath(input: unknown, result: unknown, resultPath: string | undefined): unknown {
  if (resultPath === null) return input;
  if (!resultPath || resultPath === "$") return result;
  return setPath(input, resultPath, result);
}

function applyOutputPath(data: unknown, outputPath: string | undefined): unknown {
  if (outputPath === null) return {};
  return resolvePath(data, outputPath ?? "$");
}

function applyParameters(params: Record<string, unknown> | undefined, data: unknown): unknown {
  if (!params) return data;
  const result: Record<string, unknown> = {};
  for (const [key, val] of Object.entries(params)) {
    if (key.endsWith(".$")) {
      const realKey = key.slice(0, -2);
      result[realKey] = resolvePath(data, val as string);
    } else {
      result[key] = val;
    }
  }
  return result;
}

// ─── Choice evaluator ─────────────────────────────────────────────────────────

function evaluateChoice(rule: ChoiceRule, data: unknown): boolean {
  if (rule.And) return rule.And.every((r) => evaluateChoice(r, data));
  if (rule.Or) return rule.Or.some((r) => evaluateChoice(r, data));
  if (rule.Not) return !evaluateChoice(rule.Not, data);

  const variable = rule.Variable ? resolvePath(data, rule.Variable) : undefined;

  if (rule.StringEquals !== undefined) return variable === rule.StringEquals;
  if (rule.StringLessThan !== undefined) return String(variable) < rule.StringLessThan;
  if (rule.StringGreaterThan !== undefined) return String(variable) > rule.StringGreaterThan;
  if (rule.StringLessThanOrEquals !== undefined)
    return String(variable) <= rule.StringLessThanOrEquals;
  if (rule.StringGreaterThanOrEquals !== undefined)
    return String(variable) >= rule.StringGreaterThanOrEquals;
  if (rule.StringMatches !== undefined) {
    const pattern = rule.StringMatches.replace(/[.+^${}()|[\]\\]/g, "\\$&").replace(/\*/g, ".*");
    return new RegExp(`^${pattern}$`).test(String(variable));
  }
  if (rule.NumericEquals !== undefined) return Number(variable) === rule.NumericEquals;
  if (rule.NumericLessThan !== undefined) return Number(variable) < rule.NumericLessThan;
  if (rule.NumericGreaterThan !== undefined) return Number(variable) > rule.NumericGreaterThan;
  if (rule.NumericLessThanOrEquals !== undefined)
    return Number(variable) <= rule.NumericLessThanOrEquals;
  if (rule.NumericGreaterThanOrEquals !== undefined)
    return Number(variable) >= rule.NumericGreaterThanOrEquals;
  if (rule.BooleanEquals !== undefined) return Boolean(variable) === rule.BooleanEquals;
  if (rule.IsNull !== undefined) return (variable === null) === rule.IsNull;
  if (rule.IsPresent !== undefined) return (variable !== undefined) === rule.IsPresent;
  if (rule.IsString !== undefined) return (typeof variable === "string") === rule.IsString;
  if (rule.IsNumeric !== undefined) return (typeof variable === "number") === rule.IsNumeric;
  if (rule.IsBoolean !== undefined) return (typeof variable === "boolean") === rule.IsBoolean;

  return false;
}

// ─── Task invoker ─────────────────────────────────────────────────────────────

export interface TaskInvoker {
  invoke(resource: string, input: unknown): Promise<unknown>;
}

// Default task invoker: no-op (returns input)
const defaultInvoker: TaskInvoker = {
  async invoke(_resource: string, input: unknown): Promise<unknown> {
    return input;
  },
};

// ─── Execution engine ─────────────────────────────────────────────────────────

export async function runStateMachine(
  definition: StateMachineDefinition,
  input: unknown,
  executionArn: string,
  invoker: TaskInvoker = defaultInvoker,
): Promise<{ output: unknown; error?: string; cause?: string }> {
  let currentStateName = definition.StartAt;
  let currentData = input;

  while (currentStateName) {
    const state = definition.States[currentStateName];
    if (!state) throw new StatesError("States.Runtime", `State not found: ${currentStateName}`);

    try {
      const result = await executeState(state, currentData, definition, invoker);
      if (result.terminal) {
        if (result.error) {
          return { output: input, error: result.error, cause: result.cause };
        }
        return { output: result.output };
      }
      currentData = result.output;
      currentStateName = result.next!;
    } catch (err) {
      if (err instanceof StatesError) {
        return { output: input, error: err.errorCode, cause: err.cause };
      }
      return { output: input, error: "States.Runtime", cause: String(err) };
    }
  }

  return { output: currentData };
}

interface StateResult {
  output: unknown;
  next?: string;
  terminal: boolean;
  error?: string;
  cause?: string;
}

async function executeState(
  state: AslState,
  data: unknown,
  definition: StateMachineDefinition,
  invoker: TaskInvoker,
): Promise<StateResult> {
  switch (state.Type) {
    case "Pass":
      return executePass(state as PassState, data);
    case "Task":
      return executeTask(state as TaskState, data, invoker);
    case "Choice":
      return executeChoice(state as ChoiceState, data);
    case "Wait":
      return executeWait(state as WaitState, data);
    case "Succeed":
      return executeSucceed(state as SucceedState, data);
    case "Fail":
      return executeFail(state as FailState);
    case "Parallel":
      return executeParallel(state as ParallelState, data, invoker);
    case "Map":
      return executeMap(state as MapState, data, invoker);
    default:
      throw new StatesError("States.Runtime", `Unknown state type: ${(state as AslState).Type}`);
  }
}

function executePass(state: PassState, data: unknown): StateResult {
  const effectiveInput = applyInputPath(data, state.InputPath);
  const params = applyParameters(state.Parameters, effectiveInput);
  const result = state.Result ?? params;
  const combined = applyResultPath(effectiveInput, result, state.ResultPath);
  const output = applyOutputPath(combined, state.OutputPath);

  if (state.End) return { output, terminal: true };
  return { output, next: state.Next!, terminal: false };
}

async function executeTask(
  state: TaskState,
  data: unknown,
  invoker: TaskInvoker,
): Promise<StateResult> {
  const effectiveInput = applyInputPath(data, state.InputPath);
  const params = applyParameters(state.Parameters, effectiveInput);

  let result: unknown;
  try {
    result = await invoker.invoke(state.Resource, params ?? effectiveInput);
  } catch (err) {
    throw new StatesError("States.TaskFailed", String(err));
  }

  const combined = applyResultPath(effectiveInput, result, state.ResultPath);
  const output = applyOutputPath(combined, state.OutputPath);

  if (state.End) return { output, terminal: true };
  return { output, next: state.Next!, terminal: false };
}

function executeChoice(state: ChoiceState, data: unknown): StateResult {
  const effectiveInput = applyInputPath(data, state.InputPath);

  for (const rule of state.Choices) {
    if (evaluateChoice(rule, effectiveInput)) {
      const output = applyOutputPath(effectiveInput, state.OutputPath);
      return { output, next: rule.Next, terminal: false };
    }
  }

  if (state.Default) {
    const output = applyOutputPath(effectiveInput, state.OutputPath);
    return { output, next: state.Default, terminal: false };
  }

  throw new StatesError("States.NoChoiceMatched", "No choice rule matched and no default");
}

async function executeWait(state: WaitState, data: unknown): Promise<StateResult> {
  let waitMs = 0;
  if (state.Seconds) waitMs = Math.min(state.Seconds * 1000, 5000); // cap at 5s for local
  if (state.SecondsPath) {
    const secs = resolvePath(data, state.SecondsPath);
    waitMs = Math.min(Number(secs) * 1000, 5000);
  }
  if (waitMs > 0) await new Promise((r) => setTimeout(r, waitMs));

  const effectiveInput = applyInputPath(data, state.InputPath);
  const output = applyOutputPath(effectiveInput, state.OutputPath);

  if (state.End) return { output, terminal: true };
  return { output, next: state.Next!, terminal: false };
}

function executeSucceed(state: SucceedState, data: unknown): StateResult {
  const effectiveInput = applyInputPath(data, state.InputPath);
  const output = applyOutputPath(effectiveInput, state.OutputPath);
  return { output, terminal: true };
}

function executeFail(state: FailState): StateResult {
  return {
    output: null,
    terminal: true,
    error: state.Error ?? "States.Failed",
    cause: state.Cause,
  };
}

async function executeParallel(
  state: ParallelState,
  data: unknown,
  invoker: TaskInvoker,
): Promise<StateResult> {
  const effectiveInput = applyInputPath(data, state.InputPath);

  const results = await Promise.all(
    state.Branches.map((branch) =>
      runStateMachine(
        { StartAt: branch.StartAt, States: branch.States },
        effectiveInput,
        uuidv4(),
        invoker,
      ),
    ),
  );

  const outputs = results.map((r) => r.output);
  const combined = applyResultPath(effectiveInput, outputs, state.ResultPath);
  const output = applyOutputPath(combined, state.OutputPath);

  if (state.End) return { output, terminal: true };
  return { output, next: state.Next!, terminal: false };
}

async function executeMap(
  state: MapState,
  data: unknown,
  invoker: TaskInvoker,
): Promise<StateResult> {
  const effectiveInput = applyInputPath(data, state.InputPath);
  const items = state.ItemsPath ? resolvePath(effectiveInput, state.ItemsPath) : effectiveInput;
  const itemArray = Array.isArray(items) ? items : [];

  const results = await Promise.all(
    itemArray.map((item) =>
      runStateMachine(
        { StartAt: state.Iterator.StartAt, States: state.Iterator.States },
        item,
        uuidv4(),
        invoker,
      ),
    ),
  );

  const outputs = results.map((r) => r.output);
  const combined = applyResultPath(effectiveInput, outputs, state.ResultPath);
  const output = applyOutputPath(combined, state.OutputPath);

  if (state.End) return { output, terminal: true };
  return { output, next: state.Next!, terminal: false };
}
